pkg/sql/sqlstats/insights/integration/integration_test: TestInsightsPriorityIntegration failed

pkg/sql/sqlstats/insights/integration/integration_test.TestInsightsPriorityIntegration [failed](https://mesolite.cluster.engflow.com/invocations/default/554ef68c-1a34-4249-a5e7-1737f41feea5?testReportRun=2&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3FsL3NxbHN0YXRzL2luc2lnaHRzL2ludGVncmF0aW9uOmludGVncmF0aW9uX3Rlc3Q=) on release-25.4.1-rc @ [cdee83d12f6028cc8c2dd472e0df65d7bd74d72e](https://github.com/cockroachdb/cockroach/commits/cdee83d12f6028cc8c2dd472e0df65d7bd74d72e):


```
=== RUN   TestInsightsPriorityIntegration
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestInsightsPriorityIntegration2506335065
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:154: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    insights_test.go:746: condition failed to evaluate within 3m45s: sql: no rows in result set
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestInsightsPriorityIntegration2506335065
--- FAIL: TestInsightsPriorityIntegration (240.07s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/obs-prs
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestInsightsPriorityIntegration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-57000

--- a/pkg/sql/sqlstats/insights/integration/insights_test.go
+++ b/pkg/sql/sqlstats/insights/integration/insights_test.go
@@ -726,24 +726,41 @@ func TestInsightsPriorityIntegration(t *testing.T) {
 	}
 
 	for _, p := range priorities {
-		testutils.SucceedsSoon(t, func() error {
+		// We combine running the transaction and querying for the insight into a
+		// single retry loop. This handles the case where the transaction
+		// experiences accidental contention, causing the insight to be filtered
+		// out (contention insights require resolved blocking txn fingerprints).
+		// By retrying both together, we get a fresh transaction that hopefully
+		// won't experience contention.
+		testutils.SucceedsWithin(t, func() error {
 			tx, errTxn := conn.BeginTx(ctx, &gosql.TxOptions{})
-			require.NoError(t, errTxn)
+			if errTxn != nil {
+				return errTxn
+			}
 
 			_, errTxn = tx.ExecContext(ctx, p.setPriorityQuery)
-			require.NoError(t, errTxn)
+			if errTxn != nil {
+				_ = tx.Rollback()
+				return errTxn
+			}
 
 			_, errTxn = tx.ExecContext(ctx, p.query)
-			require.NoError(t, errTxn)
+			if errTxn != nil {
+				_ = tx.Rollback()
+				return errTxn
+			}
 
 			_, errTxn = tx.ExecContext(ctx, "select pg_sleep(.1);")
-			require.NoError(t, errTxn)
-			errTxn = tx.Commit()
-			require.NoError(t, errTxn)
-			return nil
-		})
+			if errTxn != nil {
+				_ = tx.Rollback()
+				return errTxn
+			}
 
-		testutils.SucceedsSoon(t, func() error {
+			errTxn = tx.Commit()
+			if errTxn != nil {
+				return errTxn
+			}
+
 			row := conn.QueryRowContext(ctx, "SELECT "+
 				"query, "+
 				"priority, "+
@@ -772,7 +789,7 @@ func TestInsightsPriorityIntegration(t *testing.T) {
 			}
 
 			return nil
-		})
+		}, time.Minute)
 	}
 }
 

